File: Earthshake/EarthshakeBaseViewController.m
//
//  EarthshakeBaseViewController.m
//  Earthshake
//
//  Created by Andrey Paunov on 2016-04-24.
//  Copyright © 2016 AP. All rights reserved.
//

#import "EarthshakeBaseViewController.h"

@interface EarthshakeBaseViewController ()

@property (strong, nonatomic) id<EarthshakeService> earthshakeService; // Protocal to interact with the request service
@property (strong, nonatomic) NSMutableDictionary *parameters;         // Dictionary with parameters

@end

@implementation EarthshakeBaseViewController

- (void)viewDidLoad
{
    // Adding the refresh button
    UIButton *refreshBarButton = [UIButton buttonWithType: UIButtonTypeCustom];
    refreshBarButton.frame = CGRectMake(0, 0, 22, 22);
    [refreshBarButton addTarget:self action:@selector(didSelectRefresh) forControlEvents: UIControlEventTouchUpInside];
    [refreshBarButton setBackgroundImage: [UIImage imageNamed: @"Refresh"] forState: UIControlStateNormal];

    UIBarButtonItem *refreshButtonItem = [[UIBarButtonItem alloc] initWithCustomView: refreshBarButton];
    self.navigationItem.rightBarButtonItems = @[refreshButtonItem];

    // Date formatting
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"yyyy-MM-dd"];

    NSDateComponents *dateComponents = [[NSDateComponents alloc] init];
    [dateComponents setDay: -7];

    // Magnitude formatting
    self.decimalFormatter = [[NSNumberFormatter alloc] init];
    self.decimalFormatter.numberStyle = NSNumberFormatterDecimalStyle;
    self.decimalFormatter.maximumFractionDigits = 2;
    self.decimalFormatter.roundingMode = NSNumberFormatterRoundHalfUp;

    NSCalendar *calender = [NSCalendar currentCalendar];
    NSDate *endTime = [NSDate date];
    NSDate *startTime = [calender dateByAddingComponents:dateComponents toDate:endTime options:0];

    self.parameters = [[NSMutableDictionary alloc] init];
    [self.parameters setObject: [dateFormatter stringFromDate:startTime] forKey:kStartTime];
    [self.parameters setObject: [dateFormatter stringFromDate:endTime] forKey:kEndTime];
    [self.parameters setObject: MIN_MAG_VALUE forKey:kMinMagnitude];

    self.earthshakeService = [(EarthshakeAppDelegate *)[[UIApplication sharedApplication] delegate] earthshakeService];
}

#pragma mark - Helper methods

- (void)getRequestDataSuccess:(EarthshakeSuccessBlock)success failure:(EarthshakeFailureBlock)failure
{
    [self.earthshakeService getRequestData:self.parameters success:success failure:failure];
}

- (void)didSelectRefresh
{
}

- (void)displayAlert:(NSError *)error
{
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Error", nil)
                                                                   message:error.localizedDescription
                                                            preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction *dismissAction = [UIAlertAction actionWithTitle:NSLocalizedString(@"OK", nil)
                                                            style:UIAlertActionStyleDefault
                                                          handler:^(UIAlertAction * _Nonnull action)
                                    {
                                    }];
    
    [alert addAction:dismissAction];

    [self presentViewController:alert animated:YES completion:nil];
}

@end

File: Earthshake/EarthshakeBaseViewController.swift
//
//  EarthshakeBaseViewController.swift
//  Earthshake
//
//  Created by Andrey on 2016-12-30.
//  Copyright © 2016 AP. All rights reserved.
//

import Foundation

extension Selector
{
    static let refreshButtonTapped = #selector(EarthshakeBaseViewController.didSelectRefresh)
}

class EarthshakeBaseViewController : UIViewController
{
    // Earthshake GeoJson Parameters
    let kStartTime = "starttime"
    let kEndTime = "endtime"
    let kMinMagnitude = "minmagnitude"
    let MIN_MAG_VALUE = "2.5"
    let decimalFormatter = NumberFormatter()

    var refreshButton: UIButton?                // Refresh button with functionality implemented by the subclesses
    var earthshakeService: EarthshakeService?   // Protocol to interact with the request service
    var parameters = [AnyHashable: Any]()       // Dictionary with parameters

    override func viewDidLoad()
    {
        // Adding the refresh button
        refreshButton = UIButton(type: UIButtonType.custom)
        refreshButton?.frame = CGRect(x: 0, y: 0, width: 22, height: 22)
        refreshButton?.addTarget(self, action: .refreshButtonTapped, for: UIControlEvents.touchUpInside)
        refreshButton?.setBackgroundImage(UIImage(named: "Refresh"), for: UIControlState.normal)

        let refreshButtonItem = UIBarButtonItem(customView: refreshButton!)
        navigationItem.rightBarButtonItems = [refreshButtonItem]

        // Date formatting
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"

        var dateComponents = DateComponents()
        dateComponents.day = -7

        // Magnitude formatting
        decimalFormatter.numberStyle = NumberFormatter.Style.decimal
        decimalFormatter.maximumFractionDigits = 2;

        let calender = NSCalendar.current
        let endTime = Date()
        let startTime = calender.date(byAdding: dateComponents, to: endTime)

        parameters = [AnyHashable: Any]()
        parameters[kStartTime] = dateFormatter.string(from: startTime!)
        parameters[kEndTime] = dateFormatter.string(from: endTime)
        parameters[kMinMagnitude] = MIN_MAG_VALUE

        earthshakeService = ((UIApplication.shared.delegate) as! EarthshakeAppDelegate).earthshakeService
    }

    // Helper methods

    func getRequestedData(success: EarthshakeSuccessBlock, failure: EarthshakeFailureBlock)
    {
        earthshakeService?.getRequestedData(parameters, success:success, failure:failure)
    }

    func didSelectRefresh() {}

    func displayAlert(error: Error)
    {
        let alert = UIAlertController.init(title: "Error".localized, message: error.localizedDescription, preferredStyle: UIAlertControllerStyle.alert)
        let dismissAction = UIAlertAction.init(title: "OK".localized, style: UIAlertActionStyle.default) {(action) in}

        alert.addAction(dismissAction)

        present(alert, animated: true) {}
    }
}
